src/lib/Parser.hs

```diff
@@ -267,34 +268,49 @@ addImplicitImplicitArgs sourcePos (Just typ) ex =
 interfaceDef :: Parser [UDecl]
 interfaceDef = do
   keyWord InterfaceKW
-  -- TODO: allow multi-parameter type classes
-  interfaceName <- upperName <|> symName
-  (oneAnnBinder, pos) <- withPos annBinder
+  (tyCon, pos) <- withPos tyConDef
   keyWord WhereKW
-  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
-  let consName = mkInterfaceConsName interfaceName
-  let (varTypeName, oneAnnBinder') = case oneAnnBinder of
-        Bind v    -> (varName v, oneAnnBinder)
-        Ignore ty -> let genVar = mkName $ mkNoShadowingStr "gen" in
-            (genVar, Bind (genVar:>ty))
-  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder']
-  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
-  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  recordFields <- withSrc $ interfaceRecordFields ":"
+  let (UConDef interfaceName uAnnBinderNest) = tyCon
+      record = (URecordTy . NoExt) <$> recordFields
+      consName = mkInterfaceConsName interfaceName
+      ((_, varNames), uAnnBinderNest') =
+          mapAccumR mkFreeVarTypeName (0, []) uAnnBinderNest
+      tyCon' = UConDef interfaceName uAnnBinderNest'
+      funDefs = mkFunDefs (pos, varNames, interfaceName) recordFields
+  return $ (UData tyCon' [UConDef consName (toNest [Ignore record])]) : funDefs
   where
-    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
-    mkFunDefs meta ~(WithSrc _ (URecordTy (NoExt (LabeledItems items)))) =
+    mkFreeVarTypeName
+      :: (Int, [Name]) -> UAnnBinder -> ((Int, [Name]), UAnnBinder)
+    mkFreeVarTypeName (n, varNames) (Ignore ty) =
+      let genVarName = mkName $ mkNoShadowingStr ("gen" ++ show n) in
+      ((n + 1, genVarName:varNames), Bind (genVarName:>ty))
+    mkFreeVarTypeName (n, varNames) b@(Bind v) = ((n, (varName v):varNames), b)
```

Comment:
I'm trying to understand why this is necessary. So it looks like you are walking the UConDef and substituting `Ignore` binders with `Bind` binders with generated names. Then you are passing those same names down to `mkFunDefs`? But it seems to me like you don't need the names to actually be the same in the `UData` data con definition and in the functions, especially if they are ignored in the UData definition.

Could you just use `uAnnBinderNest` directly, leaving the ignore variables the same? Then when making the function definition annotations, I think you should be able to just use `UHole` for any of the ignore variables. The reasoning being, since the user didn't bind those types to a name themselves, they can't possibly be using that type anywhere in the function.

(Hm, although maybe I have that backwards and instead of being a `UHole` it should be a fresh variable. But in any case I don't think it has to be the SAME fresh variable as was used in the type definition, so I think you are free to just use the `uAnnBinderNest` in the data type definition.)

Reply:
I am not 100% sure if having a `UHole` works out, but indeed, it probably does not have to be the same variable since it can not be used; I'll check and make a change :)

Reply:
Gave it a try, and it seems that:
(1) it is not possible to use `uAnnBinderNest` directly right now (at the type checking phase we have
```hs
DataCon def@(DataDef _ paramBs cons) params con args -> do
      let paramVars = fmap (\(Bind v) -> v) paramBs
```
which relies on having only bound variables;
(2) we indeed need fresh variable names in `mkUApp`. They can indeed be different than the ones generated when building `uAnnBinderNest'`, but I am not sure how interesting/useful it is to generate different fresh variables since we have to generate `uAnnBinderNest'` either way?

Reply:
Huh. Sure enough. But I think this is actually a bug! It happens even without using interfaces:
```
>=> data Foo _:Type _:Type = MkFoo Int
dex: src/lib/Type.hs:111:29-42: Non-exhaustive patterns in lambda
```
Come to think of it, defining a TyCon with an ignored type parameter is a bit pointless. We could just forbid that in general and modify `tyConDef` to always produce `Bind`s, then this whole problem would go away. Or, we could change the typechecker to allow this kind of thing.

Either way I don't think it's worth complicating the implementation of interfaces to support an edge case that doesn't seem very useful and that we already fail to support correctly elsewhere. What do you think?

Reply:
I think it's fine to disallow it for now by modifying `tyConDef` to always produce `Bind`s, and possibly revisit the issue if we figure out a use case (in which case the typechecker will need some modification).

Does that sound good to you?

Reply:
I don't agree that types with ignored type parameters are pointless. This is precisely what a phantom type is, and they're used all over the place in Haskell. We should definitely allow that in the type checker.

Reply:
In particular, whenever it is valid to give the parameter a name that later ends up being completely unused, it should also be ok to specify the name to be `_`